### PR TITLE
Add simple plotting tool

### LINKED ISSUE #148: Speed up plotting
Plotting many realizations is a very slow process. A possible solution could be to:

- Dump to parquet in each realization
- Make a simple command line tool that plots the desired results.

The functionality should be limited - in no way do we want to compete with Webviz/ResInsight etc. We merely need a _pocket knife_ for testing purposes.

### PR DESCRIPTION
**Pull request by wouterjdb**
This PR adds a simple plotting tool which allows you to plot the results of FlowNet runs.

---

### Contributor checklist

- [x] :tada: This PR closes #148 
- [x] :scroll: I have broken down my PR into the following tasks:
   - [x] Add initial version of the tool
   - [x] Make more updates
   - [x] Add observations
   - [x] Add support for multiple vertical lines
- [ ] :robot: I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR.
- [x] :book: I have considered adding a new entry in `CHANGELOG.md`.
- [ ] :books: I have considered updating the documentation.

### CODE REVIEW

File: src/flownet/utils/plot_results.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        default=[""],
+        nargs="+",
+        help="One or #vectors unit labels.",
+    )
+    parser.add_argument(
+        "-prior_alphas",
+        type=float,
+        default=[0.1],
+        nargs="+",
+        help="One or #prior ensembles alpha (transparency) values.",
```
Comment (LonnekevB) [inline src/flownet/utils/plot_results.py:284]:
```suggestion
        help="Transparency of prior, value between 0 (transparent) and 1(opaque). Can be one number or multiple values, depending on the number of priors you are plotting. In the latter case, the number of alpha values should be equal to the number of priors. ",
```

File: src/flownet/utils/plot_results.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    )
+    parser.add_argument(
+        "-posterior_alphas",
+        type=float,
+        default=[0.1],
+        nargs="+",
+        help="One or #posterior ensembles alpha (transparency) values.",
```
Comment (LonnekevB) [inline src/flownet/utils/plot_results.py:291]:
```suggestion
        help="Transparency of posterior, value between 0 (transparent) and 1(opaque). Can be one number or multiple values, depending on the number of posteriors you are plotting. In the latter case, the number of alpha values should be equal to the number of posteriors. ",
```

File: src/flownet/utils/plot_results.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    )
+    parser.add_argument(
+        "-prior_colors",
+        type=str,
+        default=["gray"],
+        nargs="+",
+        help="One or #prior ensembles colors.",
```
Comment (LonnekevB) [inline src/flownet/utils/plot_results.py:298]:
```suggestion
        help="Color of prior lines. Can be one number or multiple colors, depending on the number of priors you are plotting. In the latter case, the number of colors should be equal to the number of priors. ",
```

File: src/flownet/utils/plot_results.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    )
+    parser.add_argument(
+        "-posterior_colors",
+        type=str,
+        default=["blue"],
+        nargs="+",
+        help="One or #posterior ensembles colors.",
```
Comment (LonnekevB) [inline src/flownet/utils/plot_results.py:305]:
```suggestion
        help="Color of posterior. Can be one number or colors, depending on the number of posteriors you are plotting. In the latter case, the number of colors should be equal to the number of posteriors. ",
```

**Review by LonnekevB — CHANGES_REQUESTED** [review]
Just some suggestions on making the help a bit more explicit.

**Review by LonnekevB — APPROVED** [review]

File: src/flownet/utils/plot_results.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+        raise ValueError(
+            f"You should either supply a single maximum y-value or as many as you have vectors ({len(args.vectors)}."
+        )
+
+    if not (len(args.units) == 1 or len(args.units) == len(args.vectors)):
+        raise ValueError(
+            f"You should either supply a single units label or as many as you have vectors ({len(args.vectors)}."
+        )
+
+    if (
```
Comment (olelod) [inline src/flownet/utils/plot_results.py:165]: If prior or posterior is not defined you will run into trouble here, I think? len(None) is not defined. Better with `if not args.prior and not args.posterior and not args.reference_simulation`?

File: src/flownet/utils/plot_results.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+        default=None,
+        help="Path to an ERT observation file.",
+    )
+    args = parser.parse_args()
+
+    check_args(args)
+
+    prior_data = build_ensemble_df_list(args.prior, args.vectors)
+    posterior_data = build_ensemble_df_list(args.posterior, args.vectors)
+
```
Comment (olelod) [inline src/flownet/utils/plot_results.py:352]: Same here. You can end up trying to iterate over `None` inside `build_ensemble_df_list`.
```suggestion
prior_data = build_ensemble_df_list(args.prior, args.vectors) if args.prior else None
posterior_data = build_ensemble_df_list(args.posterior, args.vectors) if args.posterior else None
```

File: src/flownet/utils/plot_results.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+        vector: Name of the vector to plot.
+        prior_data: List of prior ensemble data DataFrames.
+        posterior_data: List of posterior ensemble data DataFrames.
+        reference_simulation: EclSum object for the reference simulation.
+        plot_settings: Settings dictionary for the plots.
+
+    """
+    plt.figure()  # (figsize=[16, 8])
+
+    if len(prior_data):
```
Comment (olelod) [inline src/flownet/utils/plot_results.py:90]:
```suggestion
    if prior_data:
```

File: src/flownet/utils/plot_results.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+        plot_ensembles("prior", vector, prior_data, plot_settings)
+
+    if len(posterior_data):
```
Comment (olelod) [inline src/flownet/utils/plot_results.py:93]:
```suggestion
    if posterior_data:
```

**Review by olelod — CHANGES_REQUESTED** [review]

File: src/flownet/utils/plot_results.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+
+
+def plot(
+    vector: str,
+    prior_data: list,
+    posterior_data: list,
+    reference_simulation: Optional[EclSum],
+    plot_settings: dict,
+):
+    """Main plotting function that generate builds up a single plot build up
```
Comment (olelod) [inline src/flownet/utils/plot_results.py:77]: A bit hard to read... :)

```suggestion
    """Main plotting function that generate a single plot from potentially multiple ensembles and other data
```

or something

**Comment by wouterjdb** [comment]
Example from @edubarrosTNO 

image

File: src/flownet/utils/plot_results.py (hunk 6 above)
Comment (wouterjdb) [reply src/flownet/utils/plot_results.py:352]: Fixed by checking in build_ensemble_df_list if the input is None, if so I return an empty list (to make sure I at least return a list which is expected later on).

File: src/flownet/utils/plot_results.py (hunk 5 above)
Comment (wouterjdb) [reply src/flownet/utils/plot_results.py:165]: Fixed with 

```
    if (
        not args.prior is None
        and args.posterior is None
        and args.reference_simulation is None
    ):
```

And given all those parameters a default of `None` in argparse.

File: src/flownet/utils/plot_results.py (hunk 7 above)
Comment (wouterjdb) [reply src/flownet/utils/plot_results.py:90]: Now always a list and should therefore nicely evaluate to True/False. But your shorter way works just as well, so I have changed it.

File: src/flownet/utils/plot_results.py (hunk 8 above)
Comment (wouterjdb) [reply src/flownet/utils/plot_results.py:93]: Now always a list and should therefore nicely evaluate to True/False. But your shorter way works just as well, so I have changed it.

**Review by olelod — APPROVED** [review]
LGTM!

**Pull request merged**